packages/devtools_app/lib/src/memory/flutter/memory_chart.dart

```diff
@@ -167,61 +168,44 @@ class MemoryChartState extends State<MemoryChart> {
   // Trace #3 External Memory used.
   LineDataSet externalMemorySet;
 
-  // Index into the raw data.
-  int timerDataIndex;
-
-  // Active timer running.
-  Timer _timer;
-
-  Future<void> startFeed() async {
-    // Fetch from the beginning of the canned data for the live feed.
-    timerDataIndex = 0;
-
-    // TODO(terry): Consider moving the timer to the MemoryController.
-    // TODO(terry): The chart should be notified when new data arrives from the controller
-    //              using a notifier pattern.
-    // Average VMSerice rate is ~500-600 ms?
-    _timer = Timer.periodic(const Duration(milliseconds: 400), (Timer timer) {
-      if (timerDataIndex == 0) {
-        // First time reset our plotted data.
-        setState(() {
-          _used.clear();
-          _capacity.clear();
-          _externalHeap.clear();
-        });
-      }
-
-      // Pause pressed stop pumping out data simulating a live feed.
-      if (!widget.memoryController.paused &&
-          timerDataIndex < externalMemoryData.length) {
-        int x;
-        int y;
-
-        x = externalMemoryData[timerDataIndex];
-        y = externalMemoryData[timerDataIndex + 1];
-        final externalEntry =
-            Entry(x: x.toDouble(), y: y.toDouble(), icon: _img);
-
-        x = usedHeapData[timerDataIndex];
-        y = usedHeapData[timerDataIndex + 1] +
-            externalMemoryData[timerDataIndex + 1];
-        final usedEntry = Entry(x: x.toDouble(), y: y.toDouble(), icon: _img);
-
-        x = heapCapacityData[timerDataIndex];
-        y = heapCapacityData[timerDataIndex + 1];
-        final capacityEntry =
-            Entry(x: x.toDouble(), y: y.toDouble(), icon: _img);
-
-        _externalHeap.add(externalEntry);
+  void processLiveData() {
+    final MemoryController memoryController = Controllers.of(context).memory;
+
+    final List<HeapSample> liveFeed = memoryController.memoryTimeline.data;
+    if (_used.length != liveFeed.length) {
+      for (var feedIndex = _used.length;
+          feedIndex < liveFeed.length;
+          feedIndex++) {
+        final sample = liveFeed[feedIndex];
+        final timestamp = sample.timestamp.toDouble();
+
+        final capacity = sample.capacity.toDouble();
+        final used = sample.used.toDouble();
+        final external = sample.external.toDouble();
+
+        final extEntry = Entry(
+          x: timestamp,
+          y: external,
+          icon: _img,
+        );
+        final usedEntry = Entry(
+          x: timestamp,
+          y: used + external,
+          icon: _img,
+        );
+        final capacityEntry = Entry(
+          x: timestamp,
+          y: capacity,
+          icon: _img,
+        );
+
+        _externalHeap.add(extEntry);
```

Comment:
Can _externalHeap, _used, and _capacity be stored in the controller? The pattern of keeping data management to the controller has been working well for other views in devtools.

Reply:
I think your right after to talking to Dave it would be more efficient when switching between memory views and other views to not have to recompute the charting data.  I'll add a TODO to do just that.